### PR TITLE
Tree-traversal iterative design & analytic-mode

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
The tree-traversal algorithm was first implemented using a recursive approach for simplicity, but this potentially requires very deep stack calls for circuits with a lot of MCMs. Tree-traversal could also support analytic execution, optionally with a probability cutoff to ignore branches. 

**Description of the Change:**
Implement tree-traversal using an iterative approach.
Support `shots=None`.
Some profiling and optimization.

**Benefits:**
Remove system call increasing the default recursion call limit.

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-65241]

### CODE REVIEW

File: pennylane/ops/qubit/state_preparation.py [hunk 1]
```diff
@@ -169,20 +169,20 @@ class StatePrep(StatePrepBase):
     ndim_params = (1,)
     """int: Number of dimensions per trainable parameter of the operator."""
 
-    def __init__(self, state, wires, id=None):
+    def __init__(self, state, wires, id=None, skip_norm_validation=False):
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:172]:
```suggestion
    def __init__(self, state, wires, id=None, validate_norm=True):
```
I'm fine either way, but "skip" is essentially a layer of negation, so this removes a layer of negation we have to keep in our brain.

File: pennylane/measurements/mid_measure.py [hunk 2]
```diff
@@ -516,3 +516,19 @@ def __str__(self):
 
     def __repr__(self):
         return f"MeasurementValue(wires={self.wires.tolist()})"
+
+
+def find_measured_mcms(circuit):
```
Comment (albi3ro) [inline pennylane/measurements/mid_measure.py:521]: I'm curious about the decision to place this function here.

Is this going to be part of the API for the measurements module? If so, do we need to import it and document it in `measurements/__init__.py`?

File: pennylane/ops/qubit/state_preparation.py (hunk 1 above)
Comment (vincentmr) [reply pennylane/ops/qubit/state_preparation.py:172]: I agree, I'll change it to `validate_norm`.

File: pennylane/measurements/mid_measure.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/measurements/mid_measure.py:521]: > Is this going to be part of the API for the measurements module? 

I don't think so, or at least I had not envisioned it.

I'm not really sure where to put it. Could be there since it is an MCM-related utility function. I'm thinking using it to improve `dynamic_one_shot`'s performance too, so it could be in that transform's module as well. Or just `DQ::simulate` for now ...

File: pennylane/devices/default_qubit.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ def stopping_condition(op: qml.operation.Operator) -> bool:
     if op.__class__.__name__[:3] == "Pow" and qml.operation.is_trainable(op):
         return False
 
-    return op.has_matrix
+    return isinstance(op, (Conditional, MidMeasureMP)) or op.has_matrix
```
Comment (mudit2812) [inline pennylane/devices/default_qubit.py:98]: I just noticed that `adjoint_ops` (stopping condition for decomposing when using adjoint on line 173) only verifies that the tape doesn't have `MidMeasureMP`, but not `Conditional`. Could you add that too?

Comment (vincentmr) [reply pennylane/devices/default_qubit.py:98]: Nice catch, will do.

File: pennylane/measurements/mid_measure.py [hunk 4]
```diff
@@ -516,3 +516,19 @@ def __str__(self):
...
+        for op in circuit.operations
+        if isinstance(op, MidMeasureMP) and op.postselect is not None
+    )
+    for m in circuit.measurements:
+        if isinstance(m.mv, Sequence):
+            for mv in m.mv:
+                measured_mcms = measured_mcms | set(mv.measurements)
+        elif m.mv:
+            measured_mcms = measured_mcms | set(m.mv.measurements)
+    return measured_mcms
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/measurements/mid_measure.py:534]: `measured_mcms` is initialized with the MCMs that satisfy `op.postselect is not None`, so not all MCMs.

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/measurements/mid_measure.py:534]: Ah right, I missed that. But nevertheless, based on the docstring, shouldn't the return only include MCMs used for terminal measurements? In its current state the function returns all MCMs with postselection as well as MCMs used for terminal measurements.

File: pennylane/devices/qubit/simulate.py [hunk 5]
```diff
@@ -630,39 +659,229 @@ def find_next_mcm(circuit):
...
+    if circuit.shots:
+        return results
+    if not any(m.mv for m in circuit.measurements):
+        return results
+    results = list(results)
+    new_results = []
+    mid_measurements = dict((k, qml.math.array([[v]])) for k, v in mid_measurements.items())
+    for m in circuit.measurements:
+        if m.mv:
+            new_results.append(gather_mcm(m, mid_measurements, qml.math.array([[True]])))
```
Comment (mudit2812) [inline pennylane/devices/qubit/simulate.py:695]: This makes sense, but it seems a bit hacky to use a function meant to be used with finite shots.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks @vincentmr . The implementation is pretty cool. Recursion -> iteration is one of my favourite changes to see.

However, I found it a bit hard to follow, mostly because there's so much to keep track of. I found myself going back and forth a lot to remind myself of what each variable was for. With that in mind, I've suggested some additional dev comments in key parts of the algorithm. Those, along with a little bit more white space, could make the code a bit easier to parse.

I also noticed that a lot of variable names are very finite shots oriented, even in branches meant for analytic mode. 

Another point of confusion was tapes. There are a lot of different parts of the code where new tapes are created. This also made it hard to keep track of functions. For example, the `results` arg for `insert_mcms` are the results of the tape for the current depth and branch, with only non-MCM terminal measurements in the tape, but the `circuit` arg is the original circuit, not the auxiliary circuit.

I realize that this is an extremely complex, non-trivial implementation (honestly, insanely impressive, this is definitely the most complex algorithm in the entire code base), but it would be great to tweak it a bit to make it a little bit easier to follow.

File: pennylane/devices/qubit/simulate.py [hunk 6]
```diff
@@ -65,6 +65,40 @@
...
+
+    counts: dict
+    probs: dict
+    results_0: Sequence
+    results_1: Sequence
+    states: Sequence
+
+    def __init__(self, n):
+        self.counts = [None] * n
+        self.probs = [None] * n
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:80]: > The type hint for ...

I started with list-valued dicts but ended up with lists of dicts. I'll change it.

> Also, what is n?

It is the max stack depth, which is `n_mcms` + 1. I thought it would be misleading calling it `n_mcms`, so I just used a generic name. Maybe `max_depth` would be better then. What do you think?

File: pennylane/devices/qubit/simulate.py [hunk 7]
```diff
@@ -414,166 +441,160 @@ def simulate_tree_mcm(
...
-    single_measurement = len(circuit_next.measurements) == 1
-    prng_key = execution_kwargs.pop("prng_key", None)
-    for branch, count in counts.items():
-        if op.postselect is not None and branch != op.postselect:
-            prune_mcm_samples(op, branch, mcm_active, mcm_samples)
+    # This is used by `get_final_state::apply_operation` for `Conditional` operations
+    mid_measurements = dict((k, v) for k, v in zip(mcms[1:], mcm_current[1:].tolist()))
+    # Split circuit into segments
+    circuits = split_circuit_at_mcms(circuit)
+    circuits[0] = prepend_state_prep(circuits[0], None, interface, circuit.wires)
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:459]: Yes.

File: pennylane/devices/qubit/simulate.py [hunk 8]
```diff
@@ -414,166 +441,160 @@ def simulate_tree_mcm(
...
+            # Adjust counts if `invalid_postselect`
+            if invalid_postselect:
+                if finite_shots:
+                    # Bump downstream cumulative counts before zeroing-out counts
+                    for d in range(depth + 1, n_mcms + 1):
+                        cumcounts[d] += stack.counts[depth][mcm_current[depth]]
+                    stack.counts[depth][mcm_current[depth]] = 0
+                else:
+                    stack.probs[depth][mcm_current[depth]] = 0
+            measurements = tuple()
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:523]: I'll make a note that this block generates measurements for the active edge. Then I think there is a nice symmetry where one either skips execution and gets `tuple()` or performs execution and gets `measure_final_state()`.

File: pennylane/devices/qubit/simulate.py [hunk 9]
```diff
@@ -414,166 +441,160 @@ def simulate_tree_mcm(
...
+                    and mcms[depth].postselect is not None
+                    and postselect_mode == "fill-shots"
+                ):
+                    samples = mcms[depth].postselect * qml.math.ones_like(measurements)
+                else:
+                    samples = qml.math.atleast_1d(measurements)
+                stack.counts[depth] = samples_to_counts(samples)
+                stack.probs[depth] = counts_to_probs(stack.counts[depth])
+            else:
+                stack.probs[depth] = dict((k, v) for k, v in zip([False, True], measurements))
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:560]: Yes.

File: pennylane/devices/qubit/simulate.py [hunk 10]
```diff
@@ -710,16 +935,20 @@ def _(original_measurement: ProbabilityMP, measures):  # pylint: disable=unused-
...
 @combine_measurements_core.register
 def _(original_measurement: VarianceMP, measures):  # pylint: disable=unused-argument
     """Intermediate ``VarianceMP`` measurements are in fact ``SampleMP`` measurements,
     and hence the implementation is the same as for ``SampleMP``."""
-    new_sample = tuple(qml.math.atleast_1d(m[1]) for m in measures.values())
-    return np.squeeze(np.concatenate(new_sample))
+    new_sample = tuple(
+        qml.math.atleast_1d(m[1]) for m in measures.values() if m[0] and not m[1] is tuple()
+    )
+    return qml.math.squeeze(qml.math.concatenate(new_sample))
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:951]: Indeed :(

File: pennylane/devices/qubit/simulate.py (hunk 5 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:748]: Yes, it is a bit like using the integral mean value theorem when I know what the mean value is.

File: pennylane/devices/qubit/simulate.py [hunk 11]
```diff
@@ -630,39 +659,229 @@ def find_next_mcm(circuit):
...
+def variance_transform(circuit):
+    """Replace variance measurements by expectation value measurements of both the observable and the observable square.
+
+    This is necessary since computing the variance requires the global expectation value which is not available from measurements on subtrees.
+    """
+    skip_transform = circuit.shots or not any(
+        isinstance(m, VarianceMP) for m in circuit.measurements
+    )
+    if skip_transform:
+        return circuit, lambda x: x
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:869]: This is not a transform proper, `x` is already in the correct format.

File: pennylane/devices/qubit/simulate.py [hunk 12]
```diff
@@ -414,166 +441,160 @@ def simulate_tree_mcm(
...
+    finite_shots = bool(circuit.shots)
+    mcms = tuple([None] + [op for op in circuit.operations if isinstance(op, MidMeasureMP)])
+    n_mcms = len(mcms) - 1
+    mcm_current = qml.math.zeros(n_mcms + 1, dtype=bool)
+    measured_mcms = find_post_processed_mcms(circuit)
+    measured_mcms_indices = [i for i, mcm in enumerate(mcms[1:]) if mcm in measured_mcms]
+    mcm_samples = dict(
+        (k + 1, qml.math.empty((circuit.shots.total_shots,), dtype=bool) if finite_shots else None)
+        for k in measured_mcms_indices
     )
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py]: Will do.

File: pennylane/devices/qubit/simulate.py [hunk 13]
```diff
@@ -618,7 +642,12 @@ def find_next_mcm(circuit):
     for m in circuit.measurements:
         if not m.mv:
             if isinstance(m, VarianceMP):
-                new_measurements.append(SampleMP(obs=m.obs))
+                if circuit.shots:
+                    new_measurements.append(SampleMP(obs=m.obs))
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py]: Good idea, I'll use `variance_transform` in any case.

File: pennylane/devices/qubit/simulate.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/devices/qubit/simulate.py]: Yeah `max_depth` sounds great :)

File: pennylane/devices/qubit/simulate.py [hunk 14]
```diff
@@ -414,166 +441,160 @@ def simulate_tree_mcm(
...
+            # If num_shots is non-zero, simulate the current depth circuit segment
+            if depth == 0:
+                initial_state = stack.states[0]
+            else:
+                initial_state = branch_state(stack.states[depth], mcm_current[depth], mcms[depth])
+            circtmp = qml.tape.QuantumScript(
+                circuits[depth].operations,
+                circuits[depth].measurements,
+                qml.measurements.shots.Shots(shots),
+                trainable_params=circuits[depth].trainable_params,
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/devices/qubit/simulate.py]: Well with analytic execution support, I assume that eventually we also want to support backprop. Though I agree that `trainable_params` will not be used for anything. That could change in the future though if we make this implementation more general and non-`default.qubit` specific.

File: pennylane/devices/default_qubit.py [hunk 15]
```diff
@@ -67,7 +67,7 @@ def stopping_condition(op: qml.operation.Operator) -> bool:
...
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:70]: What happens if we have `Conditional(m0, qml.Permute([2,1,0], wires=(0,1,2) )`?

Comment (vincentmr) [reply pennylane/devices/default_qubit.py:70]: I would think that the operation will remain in the tape and if `m0 == 1` then `qml.Permute([2,1,0], wires=(0,1,2) )` will be applied to the state vector. Am I missing something?

File: pennylane/devices/qubit/simulate.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/devices/qubit/simulate.py:826]: This is very local so I see why you haven't made it a transform, but it does have _almost_ the same API, so maybe it's worth still adding an `@transform` decorator to it? I'd imagine the overhead would be minimal. Let me know your thoughts.

**Review by mudit2812 — COMMENTED** [review]
Looks pretty good! Just some final comments for the implementation.

The only other thing that comes to mind is testing. There's an ongoing effort right now to rework mid-circuit measurement tests, with @obliviateandsurrender and @astralcai leading the work. They might be able to provide more context, but it might be worth adding unit tests in favour of the integration tests in `test_default_qubit_native_mcm.py`.

File: pennylane/devices/default_qubit.py (hunk 15 above)
Comment (mudit2812) [reply pennylane/devices/default_qubit.py:70]: I think Christina is correct. This function checks whether or not an operator is supported by `default.qubit`. With Christina's example, the operator will be considered supported since it's a `Conditional`, but `apply_operation` will fail since `Permute` is unsupported. So we should have a separate condition for `Conditional` that also verifies that the base operator is supported.

Comment (vincentmr) [reply pennylane/devices/default_qubit.py:70]: > Permute is unsupported

Ah, that's what I was missing. Make sense now :)

File: pennylane/devices/qubit/simulate.py [hunk 16]
```diff
@@ -415,71 +438,297 @@ def simulate_tree_mcm(
...
+    # i.e. the first two MCMs had outcome 1. The last entry isn't meaningful until we are
+    # at depth `d=3`.
+    mcm_current = qml.math.zeros(n_mcms + 1, dtype=int)
+    # `mid_measurements` maps the elements of `mcm_current` to their respective MCMs
+    # This is used by `get_final_state::apply_operation` for `Conditional` operations
+    mid_measurements = dict(zip(mcms[1:], mcm_current[1:].tolist()))
+    # Split circuit into segments
+    circuits = split_circuit_at_mcms(circuit)
+    circuits[0] = prepend_state_prep(circuits[0], None, interface, circuit.wires)
+    terminal_measurements = circuits[-1].measurements if finite_shots else circuit.measurements
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:483]: Why would we have different measurements with finite shots then the analytic case?

Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:483]: In the analytic case, MCM measurements are inserted in the measurement results during tree traversal and measurements are combined as is. In the finite shots case, MCM measurements are kept in an _MCM register_ (`mcm_samples`) which is used for post-processing (deals with post-selection and terminal measurements). It might be possible to use `insert_mcms` for finite shots too, but I think it will slow down `qml.sample` since it will trigger many concatenation operations.

File: pennylane/devices/qubit/simulate.py [hunk 17]
```diff
@@ -65,6 +65,40 @@
...
+    depth-first tree-traversal procedure for simulating dynamic circuits."""
+
+    counts: Sequence
+    probs: Sequence
+    results_0: Sequence
+    results_1: Sequence
+    states: Sequence
+
+    def __init__(self, max_depth):
+        self.counts = [None] * max_depth
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:79]: Might be simple questions, but:

1) What does `None` indicate?
2) What will be the values populated into the list?

Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:79]: 1. It's a placeholder.
2. The counts, probs or states at a given depth/node (and all parent nodes). Similarly for results, either it stores the terminal measurements of a given leaf if the depth is maximal or the aggregate measurements of a subtree if the depth is intermediate. As soon as aggregate measurements are present in both branches, they are merged and the depth is decremented.

File: pennylane/devices/qubit/simulate.py [hunk 18]
```diff
@@ -65,6 +65,40 @@
 }
 
 
+class TreeTraversalStack:
+    """This class is used to record various data used during the
+    depth-first tree-traversal procedure for simulating dynamic circuits."""
+
+    counts: Sequence
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:72]: Would this be allowed to be a tuple? Do we allow it to be any type of sequence, or specifically a mutabe sequence (list)?

Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:72]: They should be lists actually since we use item assignment.

File: pennylane/devices/qubit/simulate.py [hunk 19]
```diff
@@ -440,7 +440,8 @@ def simulate_tree_mcm(
 
     # `var` measurements cannot be aggregated on the fly as they require the global `expval`
     # variance_transform replaces `var` measurements with `expval` and `expval**2` measurements
-    circuit, variance_post_processing = variance_transform(circuit)
+    circuits, variance_post_processing = variance_transform(circuit)
+    circuit = circuits[0]
```
Comment (mudit2812) [inline pennylane/devices/qubit/simulate.py:444]: Cool python syntax below:

```suggestion
    [circuit], variance_post_processing = variance_transform(circuit)
```

**Review by mudit2812 — APPROVED** [review]
Nice! Looks good to me. Very excited to get this in 🎉 . I'm approving as the only missing piece of the puzzle is test coverage, which will block merging anyway.

File: pennylane/devices/qubit/simulate.py (hunk 19 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:444]: Thanks!

**Review by albi3ro — APPROVED** [review]
👍 Sorry about the delays in reviews. In an ideal world, we'd find a way to reduce the overall complexity of the `simulate_tree_mcm` function, but not sure its the time and effort that would be involved.

Other than that, impressive work :)
